src/feature/hs/hs_intropoint.c

```diff
@@ -202,6 +203,9 @@ handle_verified_establish_intro_cell(or_circuit_t *circ,
   hs_circuitmap_register_intro_circ_v3_relay_side(circ, &auth_key);
   /* Repurpose this circuit into an intro circuit. */
   circuit_change_purpose(TO_CIRCUIT(circ), CIRCUIT_PURPOSE_INTRO_POINT);
+  /* Initialize the INTRODUCE2 token bucket for the rate limiting. */
+  token_bucket_ctr_init(&circ->introduce2_bucket, hs_dos_get_intro2_rate(),
+                        hs_dos_get_intro2_burst(), (uint32_t) approx_time());
```

Comment:
Introduction circuits can last for 18-24 hours, so it's a bit surprising that we copy the DoS values at the ESTABLISH_INTRO, and then don't change them, even if the torrc config or consensus parameters change.

That means operators will need to restart their intro point or onion service (or wait 24 hours) to overcome a DoS.

Reply:
_GREAT_ catch. 

So this patch for now only has consensus parameters support where torrc options (for the `ESTABLISH_INTRO` parameter comes with another ticket, prop305) aren't implemented yet.

This fix required a bit more code that is a function in the HS circuitmap module to "get all intro circuits" so we can go over them and adjust the token bucket parameters.

New commit takes care of this: d5e28d0d6560577e9e75826e6b3b1c72106ea7e2

Reply:
LGTM, but I will let teor take a look as well. Not resolving yet.

Reply:
Ok, looks good to me.

Let's make sure we remember to update all the circuits when the torrc options change :-)